### DISCUSSION TITLE
Third party module for all platforms

### DISCUSSION THREAD

**Discussion by igrowing** [beeware/beeware#192] [Q&A]
Hello.

[Tutorial 7](https://github.com/beeware/beeware/blob/main/docs/tutorial/tutorial-7.rst) explains how to add 3rd party lib into project.
It works and `briefcase run -u -r` indeed updated the dependencies and runs the app on Windows.

However, `briefcase run web -u -r` fails in the runtime with:
`ModuleNotFoundError: The module 'ssl' is unvendored from the Python standard library in the Pyodide distribution, you can install it by calling: await micropip.install("ssl"). See https://pyodide.org/en/stable/usage/wasm-constraints.html for more details.`

`pip install ssl` fails. And this is strange: if SSL is a dependency of HTTPX then `pip install httpx` should install it. Isn't it?

Adding `ssl` into `requires` list in pyproject.toml does not help either.
How to update the 3rd party dependencies for other platforms like Web, Android, iOS?

Thank you.

**Comment by freakboy3742** [comment]
You've hit a quirk of the tutorial - there's a very specific reason that it's done in the order that it is. The set of dependencies that works on web is [somewhat limited](https://pyodide.org/en/stable/usage/packages-in-pyodide.html#packages-in-pyodide) - especially when it comes to [anything to do with network support](https://pyodide.org/en/stable/usage/wasm-constraints.html). You can deploy anything on PyPI that is pure python; but not if it tries to access the network stack.

So - the tutorial adds HTTPX *after* the web deployment. Admittedly, this is sleight of hand; but going into the explanation why the tutorial is done in a particular order doesn't really fit well into the tutorial :-)

However, this limitation does *not* exist for iOS or Android. Those platforms have fully functioning network stacks, so you can use httpx or any other networking library if you want - and you can update requirements with `briefcase run -u -r`.

**Comment by igrowing** [comment]
Thank you @freakboy3742 for quick response. What would be a way to use "network stack" like HTTP reqest/response in Web app? Thiś is quite important because many modern apps use REST API over the network.

And just to understand the whole dependency cycle:
- If I need to use tensorflow or pytorch in my app accross all the platforms, would the method from Tutorual 7 sufficient?
- If I need to use external executables (ffmpeg, for instance) which does not exist for mobile platforms, what would be a workaround?

Thank you!

**Reply by mhsmith** [reply]
> * If I need to use tensorflow or pytorch in my app accross all the platforms, would the method from Tutorual 7 sufficient?

Yes, although Android's versions of these packages are a couple of years out of date, and iOS doesn't have them at all yet.

> * If I need to use external executables (ffmpeg, for instance) which does not exist for mobile platforms, what would be a workaround?

For Android, there are some notes about this at https://github.com/chaquo/chaquopy/issues/143.
